src/WebJobs.Script.WebHost/Controllers/KeysController.cs

```diff
@@ -0,0 +1,158 @@
+using System.Net;
+using System.Web.Http;
+using Microsoft.Azure.WebJobs.Script.WebHost.Filters;
+using Microsoft.Azure.WebJobs.Script.WebHost.Models;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Controllers
+{
+    [AuthorizationLevel(AuthorizationLevel.Admin)]
+    public class KeysController : ApiController
+    {
+        private const string MasterKeyName = "_master";
+
+        private readonly WebScriptHostManager _scriptHostManager;
+        private readonly ISecretManager _secretManager;
+
+        public KeysController(WebScriptHostManager scriptHostManager, ISecretManager secretManager)
+        {
+            _scriptHostManager = scriptHostManager;
+            _secretManager = secretManager;
+        }
+
+        [HttpGet]
+        [Route("admin/functions/{name}/keys")]
+        public IHttpActionResult Get(string name)
+        {
+            if (!_scriptHostManager.Instance.Functions.Any(f => string.Equals(f.Name, name, StringComparison.OrdinalIgnoreCase)))
+            {
+                return NotFound();
+            }
+
+            var functionKeys = _secretManager.GetFunctionSecrets(name);
+            return GetKeysResult(functionKeys);
+        }
+
+        [HttpGet]
+        [Route("admin/host/keys")]
+        public IHttpActionResult Get()
+        {
+            var hostSecrets = _secretManager.GetHostSecrets();
+            return GetKeysResult(hostSecrets.FunctionKeys);
+        }
+
+        [HttpPost]
+        [Route("admin/functions/{name}/keys/{keyName}")]
+        public IHttpActionResult Post(string name, string keyName) => AddOrUpdateFunctionSecret(keyName, null, name);
+
+        [HttpPost]
+        [Route("admin/host/keys/{keyName}")]
+        public IHttpActionResult Post(string keyName) => AddOrUpdateFunctionSecret(keyName, null);
+
+        [HttpPut]
+        [Route("admin/functions/{name}/keys/{keyName}")]
+        public IHttpActionResult Put(string name, string keyName, Key key) => PutKey(keyName, key, name);
+
+        [HttpPut]
+        [Route("admin/host/keys/{keyName}")]
+        public IHttpActionResult Put(string keyName, Key key) => PutKey(keyName, key);
+
+        [HttpDelete]
+        [Route("admin/functions/{name}/keys/{keyName}")]
+        public IHttpActionResult Delete(string name, string keyName) => DeleteFunctionSecret(keyName, name);
+
+        [HttpDelete]
+        [Route("admin/host/keys/{keyName}")]
+        public IHttpActionResult Delete(string keyName) => DeleteFunctionSecret(keyName);
+
+        private IHttpActionResult GetKeysResult(IDictionary<string, string> keys)
+        {
+            var keysContent = new
+            {
+                keys = keys.Select(k => new { name = k.Key, value = k.Value })
+            };
+
+            var keyResponse = ApiModelUtility.CreateApiModel(keysContent, Request);
+
+            return Ok(keyResponse);
+        }
+
+        private IHttpActionResult PutKey(string keyName, Key key, string functionName = null)
+        {
```

Comment:
Oh, I misread it. This returns NotFound() if the function isn't found. I thought it was doing that if the key wasn't found.

Since you do the same check first-thing in AddOrUpdateFunctionSecret, you may not have to do that in PutKey.

---
In reply to: [82860971](https://github.com/Azure/azure-webjobs-sdk-script/pull/749#discussion_r82860971) [](ancestors = 82860971)